netperf.man
.TH netperf 1L ""
.SH NAME

netperf \- a network performance benchmark

.SH SYNOPSIS

.B netperf
[global options] -- [test specific options]

.SH DESCRIPTION
.B Netperf
is a benchmark that can be used to measure various aspects of
networking performance.
Currently, its focus is on bulk data transfer and request/response
performance using either TCP or UDP, and the Berkeley Sockets
interface. In addition, tests for DLPI, and Unix Domain
Sockets, tests for IPv6 may be conditionally compiled-in.

.SS GLOBAL OPTIONS

.TP
.B \-a sizespec
Alter the send and receive buffer alignments on the local system.
This defaults to 8 bytes.
.TP
.B \-A sizespec
As -a, but for the remote system.
.TP
.B \-c [rate]
Request CPU utilization and service demand calculations for the
local system. If the optional rate parameter is specified,
.B netperf
will use that instead of calculating the rate itself.
.TP
.B \-C [rate]
As -c, but for the remote system.
.TP
.B \-d
Increase the quantity of debugging output displayed during
a test (possibly at the expense of performance).
.TP
.B \-f GMKgmk
Change the units of measure for *_STREAM tests. Capital letters are
powers of two, lowercase are powers of ten.
.TP
.B \-F fill_file
Pre-fill the send buffers with data from the named file. This is
intended to provide a means for avoiding buffers that are filled with
data which is trivially easy to compress. A good choice for a file
that should be present on any system is this manpage - netperf.man.
Other files may be provided as part of the distribution.
.TP
.B \-h
Display a usage string, and exit.
.TP
.B \-H remote_host
Set the hostname (or IP address) of the remote system.
.TP
.B \-i max,min
Set the maximum and minimum number of iterations when trying to reach
certain confidence levels.
.TP
.B \-I lvl,[,intvl]
Specify the confidence level (either 95 or 99 - 99 is the default) and
the width of the confidence interval as a percentage (default 10)
.TP
.B \-l testlen
Specify the length of the test (default 10 seconds).
A negative value sets the number of request/response transactions,
or the number of bytes for a stream test.
.TP
.B \-n numcpus
Specify the number of CPU's in the system on those systems for which
netperf has no way to find the number of CPU's programatically. 
.TP
.B \-o sizespec
Set an offset from the alignment specified with -a.
.TP
.B \-O sizespec
As -o, but for the remote system.
.TP
.B \-p portnum
Connect to a
.C netsetver
listening on the specified port, rather than using /etc/services.
.TP
.B \-P 0|1
Show (1) or suppress (0) the test banner.
.TP
.B \-t testname
Specify the test to perform.
Valid testnames are (but not always compiled-in):
.RS
.RS
.nf
.I TCP_STREAM
.I TCP_MAERTS
.I TCP_RR
.I TCP_CRR
.I UDP_STREAM
.I UDP_RR
.I TCPIPV6_STREAM
.I TCPIPV6_RR
.I TCPIPV6_CRR
.I UDPIPV6_STREAM
.I UDPIPV6_RR
.I DLCO_STREAM
.I DLCO_RR
.I DLCL_STREAM
.I DLCL_RR
.I STREAM_STREAM
.I STREAM_RR
.I DG_STREAM
.I DG_RR
.I LOC_CPU
.I REM_CPU
.fi
.RE
.RE
.TP
.B \-v verbosity
Set the verbosity level for the test (only with -P).
.TP
.B \-V
Enable the copy-avoidance features (HP-UX 9.0 and later only).

.SS TEST SPECIFIC OPTIONS

.TP
.B \-h
Display a usage string based on the test name set with -t, and exit.

Please consult the netperf manual
.I 
Netperf: A Network Performance Benchmark 
(netperf.ps) for more information. Or you can join and mail to 
netperf-talk@netperf.org.

.SH BUGS 
There is a fairly large list of known defects and misfeatures in the
manual. If you think you have found a bug, please send email to Rick
Jones <raj@cup.hp.com>.

.SH SEE ALSO
.C netserver
.br
.I
Netperf: A Network Performance Benchmark
.br
http://www.netperf.org/

.SH AUTHORS
HP Information Networks Division - Networking Performance Team.
.br
Rick Jones	<raj@cup.hp.com>
.br
Karen Choy	HP IND
.br
Dave Shield	<daves@csc.liv.ac.uk>	(man pages)
.br
Others too numerous to mention here - see the ACKNWLDGMNTS file


Release_Notes

These are the Release Notes for the 2.0PL2 revision of netperf:

Things Changed in this release:

*) Netperf.ps is a dummy file in this patch release. If you want the
   manual, either consult the Web page, or grab an earlier 2.0
   release.

*) fixed the looper process CPU util for SGI Irix, thanks to pointers
   from folks on the net on how to do it. It should now be both
   accurate, and unobtrusive.

*) added similar looper process code for AIX - untested. Feedback
   appreciated.

*) fixed a bug in the CPU utilization code when the soaker gets zero
   CPU time. Exposed on Digital Unix where nice(39) is *really*
   effective...

*) fixed a bug where netperf -t REM_CPU would leave the soaker
   proceses behind after the test, and not cleanup the
   /tmp/netperf_cpu file. There is still a problem with trying to
   measure CPU on both sides of a loopback test.

*) started fixing bugs where sockets were left behind, precluding
   running more than ~57 iterations for a confidence interval test.
   initially pointed-out by Jim Yun. some of these may not have been
   found by the time PL2 is released. Of course, the code also limits
   the max iterations to 32, so it should be rather rare that someone
   hits this.

*) added progress indications to snapshot_script so people can know
   that the script is making forward progress. These indications go to
   stderr.

*) changed the permissions on the debug file

*) just saying "./netserver" should be enough to start the standalone
   daemon  

*) other things I've probably forgotten :)

Things Not Changed in this release:

*) USE_LOOPER and USE_PSTAT still require support for "long long"

*) Still seem to be problems with Solaris 2.[34] and gcc 2.6.3 and
   2.7.0. I am unsure if this is Solaris, gcc, or netperf. Netperf
   appears to work with the Sun compilers.



netperf.c
#ifdef NEED_MAKEFILE_EDIT
#error you must first edit and customize the makefile to your platform
#endif /* NEED_MAKEFILE_EDIT */
/*
 
	   Copyright (C) 1993-2001 Hewlett-Packard Company
                         ALL RIGHTS RESERVED.
 
  The enclosed software and documention includes copyrighted works of
  Hewlett-Packard Co. For as long as you comply with the following
  limitations, you are hereby authorized to (i) use, reproduce, and
  modify the software and documentation, and to (ii) distribute the
  software and documentation, including modifications, for
  non-commercial purposes only.
      
  1.  The enclosed software and documentation is made available at no
      charge in order to advance the general development of
      high-performance networking products.
 
  2.  You may not delete any copyright notices contained in the
      software or documentation. All hard copies, and copies in
      source code or object code form, of the software or
      documentation (including modifications) must contain at least
      one of the copyright notices.
 
  3.  The enclosed software and documentation has not been subjected
      to testing and quality control and is not a Hewlett-Packard Co.
      product. At a future time, Hewlett-Packard Co. may or may not
      offer a version of the software and documentation as a product.
  
  4.  THE SOFTWARE AND DOCUMENTATION IS PROVIDED "AS IS".
      HEWLETT-PACKARD COMPANY DOES NOT WARRANT THAT THE USE,
      REPRODUCTION, MODIFICATION OR DISTRIBUTION OF THE SOFTWARE OR
      DOCUMENTATION WILL NOT INFRINGE A THIRD PARTY'S INTELLECTUAL
      PROPERTY RIGHTS. HP DOES NOT WARRANT THAT THE SOFTWARE OR
      DOCUMENTATION IS ERROR FREE. HP DISCLAIMS ALL WARRANTIES,
      EXPRESS AND IMPLIED, WITH REGARD TO THE SOFTWARE AND THE
      DOCUMENTATION. HP SPECIFICALLY DISCLAIMS ALL WARRANTIES OF
      MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
  
  5.  HEWLETT-PACKARD COMPANY WILL NOT IN ANY EVENT BE LIABLE FOR ANY
      DIRECT, INDIRECT, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES
      (INCLUDING LOST PROFITS) RELATED TO ANY USE, REPRODUCTION,
      MODIFICATION, OR DISTRIBUTION OF THE SOFTWARE OR DOCUMENTATION.
 
*/
char	netperf_id[]="\
@(#)netperf.c (c) Copyright 1993-2003 Hewlett-Packard Company. Version 2.2pl3";

#include <stdio.h>
#include <stdlib.h>
#include <strings.h>

#ifdef WIN32
#include <winsock2.h>
#include <windows.h>
#endif /* WIN32 */

#include "netsh.h"
#include "netlib.h"
#include "nettest_bsd.h"
#ifdef DO_UNIX
#include "nettest_unix.h"
#endif /* DO_UNIX */
#ifdef DO_FORE
#include "nettest_fore.h"
#endif /* DO_FORE */
#ifdef DO_HIPPI
#include "nettest_hippi.h"
#endif /* DO_HIPPI */
#ifdef DO_XTI
#include "nettest_xti.h"
#endif /* DO_XTI */
#ifdef DO_DLPI
#include "nettest_dlpi.h"
#endif /* DO_DLPI */
#ifdef DO_IPV6
#include "nettest_ipv6.h"
#endif /* DO_IPV6 */
#ifdef DO_DNS
#include "nettest_dns.h"
#endif /* DO_DNS */

 /* this file contains the main for the netperf program. all the other */
 /* routines can be found in the file netsh.c */


int _cdecl
main(int argc, char *argv[])
{

#ifdef WIN32
	WSADATA	wsa_data ;

	/* Initialize the winsock lib ( version 2.2 ) */
	if ( WSAStartup(MAKEWORD(2,2), &wsa_data) == SOCKET_ERROR ){
		printf("WSAStartup() failed : %d\n", GetLastError()) ;
		return 1 ;
	}
#endif /* WIN32 */

netlib_init();
set_defaults();
scan_cmd_line(argc,argv);

if (debug) {
	dump_globals();
	install_signal_catchers();
}

if (debug) {
	printf("remotehost is %s and port %d\n",host_name,test_port);
	fflush(stdout);
}


establish_control(host_name,test_port);

if (strcasecmp(test_name,"TCP_STREAM") == 0) {
	send_tcp_stream(host_name);
}
else if (strcasecmp(test_name,"TCP_MAERTS") == 0) {
	send_tcp_maerts(host_name);
}
#ifdef HAVE_SENDFILE
else if (strcasecmp(test_name,"TCP_SENDFILE") == 0) {
	sendfile_tcp_stream(host_name);
}
#endif /* HAVE_SENDFILE */
else if (strcasecmp(test_name,"TCP_RR") == 0) {
	send_tcp_rr(host_name);
}
else if (strcasecmp(test_name,"TCP_CRR") == 0) {
	send_tcp_conn_rr(host_name);
}
else if (strcasecmp(test_name,"TCP_CC") == 0) {
	send_tcp_cc(host_name);
}
#ifdef DO_1644
else if (strcasecmp(test_name,"TCP_TRR") == 0) {
	send_tcp_tran_rr(host_name);
}
#endif /* DO_1644 */
#ifdef DO_NBRR
else if (strcasecmp(test_name,"TCP_NBRR") == 0) {
	send_tcp_nbrr(host_name);
}
#endif /* DO_NBRR */
else if (strcasecmp(test_name,"UDP_STREAM") == 0) {
	send_udp_stream(host_name);
}
else if (strcasecmp(test_name,"UDP_RR") == 0) {
	send_udp_rr(host_name);
}
else if (strcasecmp(test_name,"LOC_CPU") == 0) {
	loc_cpu_rate();
}
else if (strcasecmp(test_name,"REM_CPU") == 0) {
	rem_cpu_rate();
}
#ifdef DO_DLPI
else if (strcasecmp(test_name,"DLCO_RR") == 0) {
	send_dlpi_co_rr(host_name);
}
else if (strcasecmp(test_name,"DLCL_RR") == 0) {
	send_dlpi_cl_rr(host_name);
}
else if (strcasecmp(test_name,"DLCO_STREAM") == 0) {
	send_dlpi_co_stream(host_name);
}
else if (strcasecmp(test_name,"DLCL_STREAM") == 0) {
	send_dlpi_cl_stream(host_name);
}
#endif /* DO_DLPI */
#ifdef DO_UNIX
else if (strcasecmp(test_name,"STREAM_RR") == 0) {
	send_stream_rr(host_name);
}
else if (strcasecmp(test_name,"DG_RR") == 0) {
	send_dg_rr(host_name);
}
else if (strcasecmp(test_name,"STREAM_STREAM") == 0) {
	send_stream_stream(host_name);
}
else if (strcasecmp(test_name,"DG_STREAM") == 0) {
	send_dg_stream(host_name);
}
#endif /* DO_UNIX */
#ifdef DO_FORE
else if (strcasecmp(test_name,"FORE_STREAM") == 0) {
	send_fore_stream(host_name);
}
else if (strcasecmp(test_name,"FORE_RR") == 0) {
	send_fore_rr(host_name);
}
#endif /* DO_FORE */
#ifdef DO_HIPPI
else if (strcasecmp(test_name,"HIPPI_STREAM") == 0) {
	send_hippi_stream(host_name);
}
else if (strcasecmp(test_name,"HIPPI_RR") == 0) {
	send_hippi_rr(host_name);
}
#endif /* DO_HIPPI */
#ifdef DO_XTI
else if (strcasecmp(test_name,"XTI_TCP_STREAM") == 0) {
	send_xti_tcp_stream(host_name);
}
else if (strcasecmp(test_name,"XTI_TCP_RR") == 0) {
	send_xti_tcp_rr(host_name);
}
else if (strcasecmp(test_name,"XTI_UDP_STREAM") == 0) {
	send_xti_udp_stream(host_name);
}
else if (strcasecmp(test_name,"XTI_UDP_RR") == 0) {
	send_xti_udp_rr(host_name);
}
#endif /* DO_XTI */
#ifdef DO_LWP
else if (strcasecmp(test_name,"LWPSTR_STREAM") == 0) {
	send_lwpstr_stream(host_name);
}
else if (strcasecmp(test_name,"LWPSTR_RR") == 0) {
	send_lwpstr_rr(host_name);
}
else if (strcasecmp(test_name,"LWPDG_STREAM") == 0) {
	send_lwpdg_stream(host_name);
}
else if (strcasecmp(test_name,"LWPDG_RR") == 0) {
	send_lwpdg_rr(host_name);
}
#endif /* DO_LWP */
#ifdef DO_IPV6
else if (strcasecmp(test_name,"TCPIPV6_STREAM") == 0) {
	send_tcpipv6_stream(host_name);
}
else if (strcasecmp(test_name,"TCPIPV6_RR") == 0) {
	send_tcpipv6_rr(host_name);
}
else if (strcasecmp(test_name,"TCPIPV6_CRR") == 0) {
	send_tcpipv6_rr(host_name);
}
else if (strcasecmp(test_name,"UDPIPV6_STREAM") == 0) {
	send_udpipv6_stream(host_name);
}
else if (strcasecmp(test_name,"UDPIPV6_RR") == 0) {
	send_udpipv6_rr(host_name);
}
#endif /* DO_IPV6 */
#ifdef DO_DNS
else if (strcasecmp(test_name,"DNS_RR") == 0) {
	send_dns_rr(host_name);
}
#endif /* DO_DNS */
else {
	printf("The test you requested is unknown to this netperf.\n");
	printf("Please verify that you have the correct test name, \n");
	printf("and that test family has been compiled into this netperf.\n");
	exit(1);
}

shutdown_control();

#ifdef WIN32
	/* Cleanup the winsock lib */
	WSACleanup();
#endif

return(0);
}


